backends/simplego/dotgeneral_fastpath.go

```diff
@@ -0,0 +1,283 @@
+		len(lhsBatchAxes) == 1 && len(rhsBatchAxes) == 1 {
+		// Batch on first axis (0)
+		if lhsBatchAxes[0] == 0 && rhsBatchAxes[0] == 0 {
+			// Contracting: lhs axis 2 with rhs axis 1
+			if lhsContractingAxes[0] == 2 && rhsContractingAxes[0] == 1 {
+				return true
+			}
+		}
+	}
+
+	// Check for multi-batch matmul: [B1, B2, M, K] × [B1, B2, K, N]
+	if lhsRank == 4 && rhsRank == 4 &&
+		len(lhsContractingAxes) == 1 && len(rhsContractingAxes) == 1 &&
+		len(lhsBatchAxes) == 2 && len(rhsBatchAxes) == 2 {
+		// Batch on first two axes
+		if lhsBatchAxes[0] == 0 && lhsBatchAxes[1] == 1 &&
+			rhsBatchAxes[0] == 0 && rhsBatchAxes[1] == 1 {
+			// Contracting: lhs axis 3 with rhs axis 2
+			if lhsContractingAxes[0] == 3 && rhsContractingAxes[0] == 2 {
+				return true
+			}
+		}
+	}
+
+	return false
+}
+
+// isBatchCrossContractOrder checks if the tensor axes are ordered as:
+// [Batch..., Cross..., Contract...]
+//
+// This ordering means:
+//   - Batch axes come first (axis 0, 1, ...)
+//   - Cross axes come next (the "output" dimensions)
+//   - Contracting axes come last (enabling sequential memory access for dot products)
+//
+// When axes are in this order, no transpose is needed before the normalized
+// DotGeneral computation.
+func isBatchCrossContractOrder(shape shapes.Shape, contractingAxes, batchAxes []int) bool {
+	rank := shape.Rank()
+	if rank == 0 {
+		return true
+	}
+
+	// Build expected order: batch axes first, then cross, then contracting
+	expectedOrder := make([]int, 0, rank)
+	expectedOrder = append(expectedOrder, batchAxes...)
+
+	// Add cross axes (non-batch, non-contracting)
+	isContracting := make(map[int]bool)
+	isBatch := make(map[int]bool)
+	for _, a := range contractingAxes {
+		isContracting[a] = true
+	}
+	for _, a := range batchAxes {
+		isBatch[a] = true
+	}
+
+	for i := 0; i < rank; i++ {
+		if !isContracting[i] && !isBatch[i] {
+			expectedOrder = append(expectedOrder, i)
+		}
+	}
+	expectedOrder = append(expectedOrder, contractingAxes...)
+
+	// Check if expected order is 0, 1, 2, ... (natural order)
+	for i, axis := range expectedOrder {
+		if axis != i {
+			return false
+		}
+	}
+	return true
+}
+
+// DirectPathMaxContractingSize is the maximum contracting dimension size for which
+// the direct (no-transpose) path is beneficial. Beyond this size, the strided RHS
+// access pattern causes too many cache misses, and the normalized path (which
+// transposes RHS for sequential access) becomes faster despite the transpose overhead.
+//
+// This threshold was determined by benchmarking (BenchmarkDirectPathThreshold):
+//   - For [256, K] × [K, 256]: DirectPath wins at K≤128, NormalizedPath wins at K≥256
+//   - Crossover point is between K=128 and K=256
+//
+// Exception: For single-row operations (M=1), DirectPath is always faster because
+// the transpose overhead dominates when there's only one output row to compute.
+const DirectPathMaxContractingSize = 128
```

Comment:
nit: leave it private for now ? --> directPathMaxConstractingSize ?

And since we are at it, what about rename all "DirectPath" by "SmallMatMul" ? Since this is about direct dot general but only for the MatMul ordered inputs that are small ?

Same elsewhere in this file ?

Reply:
updated! constants are now `smallMatMulMax{ContractingSize,BatchSize,RhsCrossSize}`